torch/csrc/jit/interpreter.cpp

```diff
@@ -645,25 +645,47 @@ struct InterpreterStateImpl {
     bool_data(function->bool_data),
     registers(function->register_size) {
   }
-  void run(Stack & stack) {
+
+  void run(Stack& stack) {
     // std::cout << *function->graph << "\n";
     // function->dump(std::cout);
     size_t pc = current_pc;
     auto & instructions = function->instructions;
     size_t last = instructions.size();
-    while(pc < last) {
+
+    // Check if we are recovered from a suspend
+    if (current_inst) {
+      registerOutputs(*current_inst, stack);
+      current_inst = nullptr;
+    }
+
+    while (pc < last) {
         // std::cout << "executing " << pc << ": ";
         // function->dumpInstruction(std::cout, pc);
         // std::cout << "\n";
         try {
           auto & inst = instructions[pc];
           loadTensorsFromRegisters(inst.inputs, stack);
-          size_t new_pc = pc + 1 + inst.callback(stack);
-          for(int i = inst.outputs.size - 1; i >= 0; i--) {
-            int reg = get(inst.outputs,i);
-            registers[reg] = pop(stack);
-            // std::cout << "pop reg[" << reg << "];\n" << registers[reg] << "\n";
+          size_t new_pc = pc + 1;
+          try {
+            new_pc += inst.callback(stack);
+          } catch (Suspend& e) {
+            JIT_ASSERT(!e.future->completed());
+            e.future->addCallback([&](){
+              c10::global_work_queue.schedule([&]() {
+                this->run(stack);
+              });
+            });
+            current_inst = &inst;
+            current_pc = new_pc;
+            // Finish what we are waiting on.
+            // This will schedule the current thread itself back into the queue.
+            // TODO: we should throw this back to the executor.
+            e.future->wait();
```

Comment:
This wait should not be here. The continuation is already scheduled, and the call to `.wait()` on the interpreter's future will ensure that progress is continued to be made.